File: tests/Aiko.Core.Tests/UpdateManifestTests.cs
using System.Security.Cryptography;
using System.Text;
using Aiko.Core;

namespace Aiko.Core.Tests;

public class UpdateManifestTests
{
    private const string PayloadHash = "5521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa63";

    // The files openssl made. The Swift port reads the same folder, so both apps agree on the format.
    private static byte[] VectorFile(string name) => SpecCases.Bytes("update-manifest", name);

    private static string VectorKey => SpecCases.Text("update-manifest", "public.pem");

    [Fact]
    public void The_openssl_vector_verifies_and_its_payload_matches()
    {
        var result = UpdateManifest.Verify(
            VectorFile("SHA256SUMS.txt"), VectorFile("SHA256SUMS.txt.sig"), VectorKey);

        Assert.Equal(ManifestStatus.Ok, result.Status);
        Assert.Equal(["payload.bin"], result.Manifest!.FileNames);
        Assert.Equal(FileCheck.Ok, result.Manifest.CheckFile("payload.bin", VectorFile("payload.bin")));
    }

    [Fact]
    public void A_manifest_changed_by_one_byte_has_a_bad_signature()
    {
        var manifest = VectorFile("SHA256SUMS.txt");
        manifest[0] = manifest[0] == (byte)'5' ? (byte)'6' : (byte)'5';

        var result = UpdateManifest.Verify(manifest, VectorFile("SHA256SUMS.txt.sig"), VectorKey);

        Assert.Equal(ManifestStatus.BadSignature, result.Status);
        Assert.Null(result.Manifest);
    }

    [Fact]
    public void A_signature_from_another_key_is_bad()
    {
        using var otherKey = ECDsa.Create(ECCurve.NamedCurves.nistP256);
        var manifest = VectorFile("SHA256SUMS.txt");

        var result = UpdateManifest.Verify(manifest, Sign(manifest, otherKey), VectorKey);

        Assert.Equal(ManifestStatus.BadSignature, result.Status);
    }

    [Fact]
    public void A_truncated_signature_is_bad()
    {
        var signature = VectorFile("SHA256SUMS.txt.sig");

        var result = UpdateManifest.Verify(VectorFile("SHA256SUMS.txt"), signature[..^1], VectorKey);

        Assert.Equal(ManifestStatus.BadSignature, result.Status);
    }

    [Fact]
    public void An_empty_signature_is_bad()
    {
        var result = UpdateManifest.Verify(VectorFile("SHA256SUMS.txt"), [], VectorKey);

        Assert.Equal(ManifestStatus.BadSignature, result.Status);
    }

    [Fact]
    public void A_signature_in_the_raw_format_instead_of_der_is_bad()
    {
        // CryptoKit and .NET both have a raw r||s form. The release uses DER, and only DER counts.
        using var key = ECDsa.Create(ECCurve.NamedCurves.nistP256);
        var manifest = Encoding.ASCII.GetBytes($"{PayloadHash}  payload.bin\n");
        var raw = key.SignData(
            manifest, HashAlgorithmName.SHA256, DSASignatureFormat.IeeeP1363FixedFieldConcatenation);

        var result = UpdateManifest.Verify(manifest, raw, PublicPem(key));

        Assert.Equal(ManifestStatus.BadSignature, result.Status);
    }

    [Fact]
    public void A_changed_payload_does_not_match_its_line()
    {
        var manifest = Verified(VectorFile("SHA256SUMS.txt"));
        var payload = VectorFile("payload.bin").Append((byte)0).ToArray();

        Assert.Equal(FileCheck.HashMismatch, manifest.CheckFile("payload.bin", payload));
    }

    [Fact]
    public void A_file_the_manifest_does_not_list_is_not_listed()
    {
        var manifest = Verified(VectorFile("SHA256SUMS.txt"));

        Assert.Equal(FileCheck.NotListed, manifest.CheckFile("Payload.bin", VectorFile("payload.bin")));
        Assert.Equal(FileCheck.NotListed, manifest.CheckFile("other.bin", VectorFile("payload.bin")));
    }

    [Fact]
    public void A_precomputed_hash_is_checked_like_the_file()
    {
        var manifest = Verified(VectorFile("SHA256SUMS.txt"));

        Assert.Equal(FileCheck.Ok, manifest.CheckHash("payload.bin", Convert.FromHexString(PayloadHash)));
        Assert.Equal(FileCheck.HashMismatch, manifest.CheckHash("payload.bin", new byte[32]));
        Assert.Equal(FileCheck.HashMismatch, manifest.CheckHash("payload.bin", new byte[31]));
    }

    [Fact]
    public void A_manifest_with_several_files_verifies_with_a_key_made_in_the_test()
    {
        using var key = ECDsa.Create(ECCurve.NamedCurves.nistP256);
        var setupHash = Hex(SHA256.HashData("setup"u8));
        var manifest = Encoding.ASCII.GetBytes(
            $"{setupHash}  Slayumind.Aiko-win-Setup.exe\n{PayloadHash}  releases.win.json\n");

        var result = UpdateManifest.Verify(manifest, Sign(manifest, key), PublicPem(key));

        Assert.Equal(ManifestStatus.Ok, result.Status);
        Assert.Equal(FileCheck.Ok, result.Manifest!.CheckFile("Slayumind.Aiko-win-Setup.exe", "setup"u8));
        Assert.Equal(FileCheck.HashMismatch, result.Manifest.CheckFile("releases.win.json", "setup"u8));
    }

    [Fact]
    public void A_crlf_manifest_is_rejected_even_when_signed()
    {
        // The workflow writes LF. A CR would end up in every file name, so CRLF is not accepted.
        using var key = ECDsa.Create(ECCurve.NamedCurves.nistP256);
        var manifest = Encoding.ASCII.GetBytes($"{PayloadHash}  payload.bin\r\n");

        var result = UpdateManifest.Verify(manifest, Sign(manifest, key), PublicPem(key));

        Assert.Equal(ManifestStatus.Malformed, result.Status);
        Assert.Null(result.Manifest);
    }

    [Theory]
    [InlineData("")]
    [InlineData("\n")]
    [InlineData(PayloadHash + "  payload.bin")] // no LF at the end
    [InlineData(PayloadHash + "  payload.bin\n\n")] // blank line
    [InlineData(PayloadHash + " payload.bin\n")] // one space
    [InlineData(PayloadHash + " *payload.bin\n")] // binary mode marker of sha256sum
    [InlineData(PayloadHash + "  \n")] // no name
    [InlineData(PayloadHash + "   payload.bin\n")] // name starts with a space
    [InlineData(PayloadHash + "  dir/payload.bin\n")]
    [InlineData(PayloadHash + "  dir\\payload.bin\n")]
    [InlineData(PayloadHash + "  ..\n")]
    [InlineData(PayloadHash + "  pay\tload.bin\n")]
    [InlineData("5521DD4A9EEACCCDEF7DD1C7EA644FB1A1891157D5DAED2ECA8B6E47C3D8AA63  payload.bin\n")]
    [InlineData("5521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa6  payload.bin\n")]
    [InlineData("5521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa633  payload.bin\n")]
    [InlineData("x521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa63  payload.bin\n")]
    [InlineData(PayloadHash + "  payload.bin\n" + PayloadHash + "  payload.bin\n")] // duplicate
    public void Anything_but_the_workflow_format_is_malformed(string text)
    {
        Assert.False(UpdateManifest.TryParse(Encoding.UTF8.GetBytes(text), out var manifest));
        Assert.Null(manifest);
    }

    [Fact]
    public void A_name_outside_ascii_is_malformed()
    {
        var bytes = Encoding.UTF8.GetBytes($"{PayloadHash}  paylöad.bin\n");

        Assert.False(UpdateManifest.TryParse(bytes, out _));
    }

    [Fact]
    public void Names_that_differ_only_in_case_are_two_files()
    {
        var bytes = Encoding.ASCII.GetBytes($"{PayloadHash}  a.bin\n{PayloadHash}  A.bin\n");

        Assert.True(UpdateManifest.TryParse(bytes, out var manifest));
        Assert.Equal(2, manifest!.FileNames.Count);
    }

    [Fact]
    public void A_key_that_is_not_pem_is_a_bad_key()
    {
        var result = UpdateManifest.Verify(
            VectorFile("SHA256SUMS.txt"), VectorFile("SHA256SUMS.txt.sig"), "not a key");

        Assert.Equal(ManifestStatus.BadKey, result.Status);
    }

    [Fact]
    public void A_private_key_is_a_bad_key()
    {
        // The app must carry only the public half. A private key in its place is a mistake to catch.
        using var key = ECDsa.Create(ECCurve.NamedCurves.nistP256);
        var manifest = VectorFile("SHA256SUMS.txt");

        var result = UpdateManifest.Verify(manifest, Sign(manifest, key), key.ExportPkcs8PrivateKeyPem());

        Assert.Equal(ManifestStatus.BadKey, result.Status);
    }

    [Theory]
    [InlineData("nistP384")]
    [InlineData("brainpoolP256r1")] // the same size as P-256, so the size alone proves nothing
    public void A_key_on_another_curve_is_a_bad_key(string curve)
    {
        using var key = ECDsa.Create(ECCurve.CreateFromFriendlyName(curve));
        var manifest = VectorFile("SHA256SUMS.txt");

        var result = UpdateManifest.Verify(manifest, Sign(manifest, key), PublicPem(key));

        Assert.Equal(ManifestStatus.BadKey, result.Status);
    }

    [Fact]
    public void A_public_key_with_broken_base64_is_a_bad_key()
    {
        var pem = "-----BEGIN PUBLIC KEY-----\nAAAA\n-----END PUBLIC KEY-----\n";

        var result = UpdateManifest.Verify(
            VectorFile("SHA256SUMS.txt"), VectorFile("SHA256SUMS.txt.sig"), pem);

        Assert.Equal(ManifestStatus.BadKey, result.Status);
    }

    [Fact]
    public void A_comment_above_the_key_is_allowed()
    {
        var pem = "# Replace this with the real key.\n" + VectorKey;

        var result = UpdateManifest.Verify(
            VectorFile("SHA256SUMS.txt"), VectorFile("SHA256SUMS.txt.sig"), pem);

        Assert.Equal(ManifestStatus.Ok, result.Status);
    }

    [Fact]
    public void The_one_file_with_an_ending_is_found()
    {
        // How the macOS app finds its zip without knowing the name the workflow gave it.
        var manifest = Verified(VectorFile("SHA256SUMS.txt"));

        Assert.Equal("payload.bin", manifest.OnlyFileEndingWith(".bin"));
        Assert.Equal("payload.bin", manifest.OnlyFileEndingWith("payload.bin"));
        Assert.Null(manifest.OnlyFileEndingWith(".zip"));
        Assert.Null(manifest.OnlyFileEndingWith(".BIN"));
    }

    [Fact]
    public void Two_files_with_the_same_ending_are_no_answer()
    {
        // A release that changed shape. Guessing which zip is the app is how the wrong one gets
        // installed, so Aiko answers nothing at all.
        var text = $"{PayloadHash}  Aiko-mac.zip\n{PayloadHash}  Aiko-mac-debug.zip\n";

        Assert.True(UpdateManifest.TryParse(Encoding.ASCII.GetBytes(text), out var manifest));
        Assert.Null(manifest!.OnlyFileEndingWith(".zip"));
        Assert.Equal("Aiko-mac-debug.zip", manifest.OnlyFileEndingWith("debug.zip"));
    }

    private static UpdateManifest Verified(byte[] manifest)
    {
        var result = UpdateManifest.Verify(manifest, VectorFile("SHA256SUMS.txt.sig"), VectorKey);
        Assert.Equal(ManifestStatus.Ok, result.Status);
        return result.Manifest!;
    }

    private static byte[] Sign(byte[] data, ECDsa key) =>
        key.SignData(data, HashAlgorithmName.SHA256, DSASignatureFormat.Rfc3279DerSequence);

    private static string PublicPem(ECDsa key) => key.ExportSubjectPublicKeyInfoPem();

    private static string Hex(byte[] bytes) => Convert.ToHexStringLower(bytes);
}

File: mac/Tests/AikoKitTests/UpdateManifestTests.swift
import CryptoKit
import Foundation
import Testing

@testable import AikoKit

/// The twin of UpdateManifestTests.cs. The same files under spec/cases/update-manifest and the
/// same answers, so a release either passes on both systems or on neither.
struct UpdateManifestTests {
    static let payloadHash = "5521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa63"

    /// Made with `openssl ecparam -name brainpoolP256r1`. The same key size as P-256, so the size
    /// alone proves nothing; the curve has to be read.
    static let brainpoolKey = """
        -----BEGIN PUBLIC KEY-----
        MFowFAYHKoZIzj0CAQYJKyQDAwIIAQEHA0IABHFoCnvQtsAGE4GOSS0xVkSHDO44
        E6KU6BVu2e4reOJWcLf0oi2rLuDTjs/4OB7ZWopVZRGWEVJFyjpiVAT3DK8=
        -----END PUBLIC KEY-----
        """

    /// A P-384 key and a brainpool key of the same size as ours. Both are refused as keys.
    static let keysOnAnotherCurve: [String] = [
        P384.Signing.PrivateKey().publicKey.pemRepresentation,
        brainpoolKey,
    ]

    static let notTheWorkflowFormat: [String] = [
        "",
        "\n",
        payloadHash + "  payload.bin",  // no LF at the end
        payloadHash + "  payload.bin\n\n",  // blank line
        payloadHash + " payload.bin\n",  // one space
        payloadHash + " *payload.bin\n",  // binary mode marker of sha256sum
        payloadHash + "  \n",  // no name
        payloadHash + "   payload.bin\n",  // name starts with a space
        payloadHash + "  dir/payload.bin\n",
        payloadHash + "  dir\\payload.bin\n",
        payloadHash + "  ..\n",
        payloadHash + "  pay\tload.bin\n",
        "5521DD4A9EEACCCDEF7DD1C7EA644FB1A1891157D5DAED2ECA8B6E47C3D8AA63  payload.bin\n",
        "5521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa6  payload.bin\n",
        "5521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa633  payload.bin\n",
        "x521dd4a9eeacccdef7dd1c7ea644fb1a1891157d5daed2eca8b6e47c3d8aa63  payload.bin\n",
        payloadHash + "  payload.bin\n" + payloadHash + "  payload.bin\n",  // duplicate
    ]

    // The files openssl made. The Windows tests read the same folder.
    static func vectorFile(_ name: String) -> Data {
        SpecCases.bytes("update-manifest", name)
    }

    static var vectorKey: String { SpecCases.text("update-manifest", "public.pem") }

    @Test
    func theOpensslVectorVerifiesAndItsPayloadMatches() {
        let result = UpdateManifest.verify(
            Self.vectorFile("SHA256SUMS.txt"), Self.vectorFile("SHA256SUMS.txt.sig"), Self.vectorKey)

        #expect(result.status == .ok)
        #expect(result.manifest?.fileNames == ["payload.bin"])
        #expect(result.manifest?.checkFile("payload.bin", Self.vectorFile("payload.bin")) == .ok)
    }

    @Test
    func aManifestChangedByOneByteHasABadSignature() {
        var manifest = Self.vectorFile("SHA256SUMS.txt")
        manifest[manifest.startIndex] = manifest[manifest.startIndex] == UInt8(ascii: "5")
            ? UInt8(ascii: "6") : UInt8(ascii: "5")

        let result = UpdateManifest.verify(
            manifest, Self.vectorFile("SHA256SUMS.txt.sig"), Self.vectorKey)

        #expect(result.status == .badSignature)
        #expect(result.manifest == nil)
    }

    @Test
    func aSignatureFromAnotherKeyIsBad() {
        let other = P256.Signing.PrivateKey()
        let manifest = Self.vectorFile("SHA256SUMS.txt")
        let signature = try! other.signature(for: manifest).derRepresentation

        let result = UpdateManifest.verify(manifest, signature, Self.vectorKey)

        #expect(result.status == .badSignature)
    }

    @Test
    func aTruncatedSignatureIsBad() {
        let signature = Self.vectorFile("SHA256SUMS.txt.sig").dropLast()

        let result = UpdateManifest.verify(
            Self.vectorFile("SHA256SUMS.txt"), Data(signature), Self.vectorKey)

        #expect(result.status == .badSignature)
    }

    @Test
    func anEmptySignatureIsBad() {
        let result = UpdateManifest.verify(Self.vectorFile("SHA256SUMS.txt"), Data(), Self.vectorKey)

        #expect(result.status == .badSignature)
    }

    @Test
    func aSignatureInTheRawFormatInsteadOfDerIsBad() {
        // CryptoKit and .NET both have a raw r||s form. The release uses DER, and only DER counts.
        let key = P256.Signing.PrivateKey()
        let manifest = Data("\(Self.payloadHash)  payload.bin\n".utf8)
        let raw = try! key.signature(for: manifest).rawRepresentation

        let result = UpdateManifest.verify(manifest, raw, key.publicKey.pemRepresentation)

        #expect(result.status == .badSignature)
    }

    @Test
    func aChangedPayloadDoesNotMatchItsLine() {
        let manifest = Self.verified(Self.vectorFile("SHA256SUMS.txt"))
        var payload = Self.vectorFile("payload.bin")
        payload.append(0)

        #expect(manifest?.checkFile("payload.bin", payload) == .hashMismatch)
    }

    @Test
    func aFileTheManifestDoesNotListIsNotListed() {
        let manifest = Self.verified(Self.vectorFile("SHA256SUMS.txt"))

        #expect(manifest?.checkFile("Payload.bin", Self.vectorFile("payload.bin")) == .notListed)
        #expect(manifest?.checkFile("other.bin", Self.vectorFile("payload.bin")) == .notListed)
    }

    @Test
    func aPrecomputedHashIsCheckedLikeTheFile() {
        let manifest = Self.verified(Self.vectorFile("SHA256SUMS.txt"))

        #expect(manifest?.checkHash("payload.bin", Self.bytes(Self.payloadHash)) == .ok)
        #expect(manifest?.checkHash("payload.bin", [UInt8](repeating: 0, count: 32)) == .hashMismatch)
        #expect(manifest?.checkHash("payload.bin", [UInt8](repeating: 0, count: 31)) == .hashMismatch)
    }

    @Test
    func aManifestWithSeveralFilesVerifiesWithAKeyMadeInTheTest() {
        let key = P256.Signing.PrivateKey()
        let setupHash = SHA256.hash(data: Data("setup".utf8)).map { String(format: "%02x", $0) }.joined()
        let manifest = Data(
            "\(setupHash)  Slayumind.Aiko-win-Setup.exe\n\(Self.payloadHash)  releases.win.json\n".utf8)
        let signature = try! key.signature(for: manifest).derRepresentation

        let result = UpdateManifest.verify(manifest, signature, key.publicKey.pemRepresentation)

        #expect(result.status == .ok)
        #expect(result.manifest?.checkFile("Slayumind.Aiko-win-Setup.exe", Data("setup".utf8)) == .ok)
        #expect(result.manifest?.checkFile("releases.win.json", Data("setup".utf8)) == .hashMismatch)
    }

    @Test
    func aCrlfManifestIsRejectedEvenWhenSigned() {
        // The workflow writes LF. A CR would end up in every file name, so CRLF is not accepted.
        let key = P256.Signing.PrivateKey()
        let manifest = Data("\(Self.payloadHash)  payload.bin\r\n".utf8)
        let signature = try! key.signature(for: manifest).derRepresentation

        let result = UpdateManifest.verify(manifest, signature, key.publicKey.pemRepresentation)

        #expect(result.status == .malformed)
        #expect(result.manifest == nil)
    }

    @Test(arguments: UpdateManifestTests.notTheWorkflowFormat)
    func anythingButTheWorkflowFormatIsMalformed(text: String) {
        #expect(UpdateManifest.parse(Data(text.utf8)) == nil)
    }

    @Test
    func aNameOutsideAsciiIsMalformed() {
        #expect(UpdateManifest.parse(Data("\(Self.payloadHash)  paylöad.bin\n".utf8)) == nil)
    }

    @Test
    func namesThatDifferOnlyInCaseAreTwoFiles() {
        let manifest = UpdateManifest.parse(
            Data("\(Self.payloadHash)  a.bin\n\(Self.payloadHash)  A.bin\n".utf8))

        #expect(manifest?.fileNames.count == 2)
    }

    @Test
    func theOneFileWithAnEndingIsFound() {
        // How the macOS app finds its zip without knowing the name the workflow gave it.
        let manifest = Self.verified(Self.vectorFile("SHA256SUMS.txt"))

        #expect(manifest?.onlyFileEndingWith(".bin") == "payload.bin")
        #expect(manifest?.onlyFileEndingWith("payload.bin") == "payload.bin")
        #expect(manifest?.onlyFileEndingWith(".zip") == nil)
        #expect(manifest?.onlyFileEndingWith(".BIN") == nil)
    }

    @Test
    func twoFilesWithTheSameEndingAreNoAnswer() {
        // A release that changed shape. Guessing which zip is the app is how the wrong one gets
        // installed, so Aiko answers nothing at all.
        let manifest = UpdateManifest.parse(
            Data("\(Self.payloadHash)  Aiko-mac.zip\n\(Self.payloadHash)  Aiko-mac-debug.zip\n".utf8))

        #expect(manifest?.onlyFileEndingWith(".zip") == nil)
        #expect(manifest?.onlyFileEndingWith("debug.zip") == "Aiko-mac-debug.zip")
    }

    @Test
    func aKeyThatIsNotPemIsABadKey() {
        let result = UpdateManifest.verify(
            Self.vectorFile("SHA256SUMS.txt"), Self.vectorFile("SHA256SUMS.txt.sig"), "not a key")

        #expect(result.status == .badKey)
    }

    @Test
    func aPrivateKeyIsABadKey() {
        // The app must carry only the public half. A private key in its place is a mistake to catch.
        let key = P256.Signing.PrivateKey()
        let manifest = Self.vectorFile("SHA256SUMS.txt")
        let signature = try! key.signature(for: manifest).derRepresentation

        let result = UpdateManifest.verify(manifest, signature, key.pemRepresentation)

        #expect(result.status == .badKey)
    }

    @Test(arguments: UpdateManifestTests.keysOnAnotherCurve)
    func aKeyOnAnotherCurveIsABadKey(pem: String) {
        let result = UpdateManifest.verify(
            Self.vectorFile("SHA256SUMS.txt"), Self.vectorFile("SHA256SUMS.txt.sig"), pem)

        #expect(result.status == .badKey)
    }

    @Test
    func aPublicKeyWithBrokenBase64IsABadKey() {
        let pem = "-----BEGIN PUBLIC KEY-----\nAAAA\n-----END PUBLIC KEY-----\n"

        let result = UpdateManifest.verify(
            Self.vectorFile("SHA256SUMS.txt"), Self.vectorFile("SHA256SUMS.txt.sig"), pem)

        #expect(result.status == .badKey)
    }

    @Test
    func aCommentAboveTheKeyIsAllowed() {
        let pem = "# Replace this with the real key.\n" + Self.vectorKey

        let result = UpdateManifest.verify(
            Self.vectorFile("SHA256SUMS.txt"), Self.vectorFile("SHA256SUMS.txt.sig"), pem)

        #expect(result.status == .ok)
    }

    // ---- helpers ----

    static func verified(_ manifest: Data) -> UpdateManifest? {
        let result = UpdateManifest.verify(manifest, vectorFile("SHA256SUMS.txt.sig"), vectorKey)
        #expect(result.status == .ok)
        return result.manifest
    }

    static func bytes(_ hex: String) -> [UInt8] {
        var out: [UInt8] = []
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            out.append(UInt8(hex[index..<next], radix: 16)!)
            index = next
        }
        return out
    }
}
